### PR TITLE
add supportf for Socket.DontFragment on macOS

### LINKED ISSUE #76422: The DontFragment SocketOptionName doesn't appear to work on macOS
### Description

Attempting to set DontFragment on a socket using `SetSocketOption` will cause a SocketException with error code 45, making it so that you can't have a socket that doesn't send fragmented packets without an exception occurring.

### Reproduction Steps

```cs
var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
socket.Bind(IPEndPoint.Parse("0.0.0.0:6500");
socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.DontFragment, true);
```

### Expected behavior

The socket option to be correctly set.

### Actual behavior

```
Unhandled exception. System.Net.Sockets.SocketException (45): Operation not supported
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Int32 optionValue, Boolean silent)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Int32 optionValue)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Boolean optionValue)
```

### Regression?

_No response_

### Known Workarounds

None, that I know of. I cannot currently figure out how to actually set the raw option (it seems to be a boolean according to examples I've seen, but I've been having issues with passing in a read only span of bytes where it refuses to accept the byte span).

### Configuration

- NET 6.0.400
- macOS 11 (Big Sur)
- x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by wfurt**
fixes #76422
It seems like the option was added in 11.0 and we are calling support for 12+ and that is also where we run the tests. 

It works on UDP & ICMP and  it seems to be silently ignored by kernel on TCP.

image

### CODE REVIEW

File: src/native/libs/System.Native/pal_networking.c [hunk 1]
```diff
@@ -1962,7 +1966,6 @@ int32_t SystemNative_GetSockOpt(
     }
 
     int fd = ToFileDescriptor(socket);
-
```
Comment (stephentoub) [inline src/native/libs/System.Native/pal_networking.c:1965]: Nit: can you put back this line?

**Review by stephentoub — APPROVED** [review]

File: src/native/libs/System.Native/pal_networking.c [hunk 2]
```diff
@@ -1795,6 +1795,12 @@ static bool TryGetPlatformSocketOption(int32_t socketOptionLevel, int32_t socket
                     return true;
 #endif
 
+#ifdef IP_DONTFRAG
+                case SocketOptionName_SO_IP_DONTFRAGMENT:
+                    *optName = IP_DONTFRAG;
```
Comment (antonfirsov) [inline src/native/libs/System.Native/pal_networking.c:1800]: Will this work for IPv6, or does that case need to be handled by using `IPV6_DONTFRAG` instead?

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/UdpClientTest.cs [hunk 3]
```diff
@@ -264,7 +264,6 @@ public void Ttl_Roundtrips()
             }
         }
 
-        [SkipOnPlatform(TestPlatforms.OSX | TestPlatforms.FreeBSD, "BSD like doesn't have an equivalent of DontFragment")]
         [Fact]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/51392", TestPlatforms.iOS | TestPlatforms.tvOS | TestPlatforms.MacCatalyst)]
         public void DontFragment_Roundtrips()
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/UdpClientTest.cs:269]: Would be nice to extend the test to also try IPv6.

Comment (wfurt) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/UdpClientTest.cs:269]: We have Tests for IPv6 and it is expected to throw as in general IPv6 does not fragment and forces PMTU discovery. 

https://github.com/dotnet/runtime/blob/2dc8c956f98389418400dda5bda663d67cbea938/src/libraries/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs#L123

 #76410 is asking to change that but that is another works cross the platforms.  (if we decide to move forward)

**Review by antonfirsov — APPROVED** [review]

**Comment by wfurt** [comment]
all test failures are known issues

**Pull request merged**